// @upstash/workflow/express.mjs
import {
  SDK_TELEMETRY,
  serveBase,
  serveManyBase
} from "./chunk-G24J5PCC.mjs";

// platforms/express.ts
import {
  Router
} from "express";
var isEmptyRequest = (req) => {
  return req.headers["content-type"] === "application/json" && req.headers["content-length"] === "0";
};
var telemetry = {
  sdk: SDK_TELEMETRY,
  framework: "express",
  runtime: process.versions.bun ? `bun@${process.versions.bun}/node@${process.version}` : `node@${process.version}`
};
function createExpressHandler(params) {
  const [routeFunction, options] = params;
  return async (request_, res) => {
    if (request_.method.toUpperCase() !== "POST") {
      res.status(405).json("Only POST requests are allowed in workflows");
      return;
    }
    let requestBody;
    if (isEmptyRequest(request_)) {
      requestBody = "";
    } else if (request_.headers["content-type"]?.includes("text/plain")) {
      requestBody = request_.body;
    } else if (request_.headers["content-type"]?.includes("application/json")) {
      requestBody = JSON.stringify(request_.body);
    } else {
      requestBody = typeof request_.body === "string" ? request_.body : JSON.stringify(request_.body);
    }
    const protocol = request_.protocol;
    const host = request_.get("host") || "localhost";
    const url = `${protocol}://${host}${request_.originalUrl}`;
    const webRequest = new Request(url, {
      method: request_.method,
      headers: new Headers(request_.headers),
      body: requestBody
    });
    const { handler: serveHandler } = serveBase(
      routeFunction,
      telemetry,
      options,
      {
        useJSONContent: true
      }
    );
    const response = await serveHandler(webRequest);
    for (const [key, value] of response.headers.entries()) {
      res.setHeader(key, value);
    }
    const responseData = await response.json();
    res.status(response.status).json(responseData);
  };
}
function serve(routeFunction, options) {
  const router = Router();
  const handler = createExpressHandler([routeFunction, options]);
  router.all(/(.*)/, handler);
  return router;
}
var createWorkflow = (...params) => {
  const [routeFunction, options = {}] = params;
  return {
    routeFunction,
    options,
    workflowId: void 0,
    useJSONContent: true
  };
};
var serveMany = (workflows, options) => {
  const router = Router();
  const { handler } = serveManyBase({
    workflows,
    getUrl(...params) {
      return params[0].url;
    },
    serveMethod: (...params) => createExpressHandler(params),
    options
  });
  router.all(/(.*)/, handler);
  return router;
};
export {
  createWorkflow,
  serve,
  serveMany
};
