// @upstash/workflow/index.mjs
import {
  SDK_TELEMETRY,
  StepTypes,
  WORKFLOW_LABEL_HEADER,
  WorkflowAbort,
  WorkflowContext,
  WorkflowError,
  WorkflowMiddleware,
  WorkflowNonRetryableError,
  WorkflowRetryAfterError,
  buildBulkActionQueryParameters,
  getWorkflowRunId,
  loggingMiddleware,
  makeGetWaitersRequest,
  makeNotifyRequest,
  normalizeCursor,
  prepareFlowControl,
  serve,
  triggerFirstInvocation
} from "./chunk-G24J5PCC.mjs";

// src/client/index.ts
import { Client as QStashClient } from "@upstash/qstash";

// src/client/dlq.ts
function buildResumeRestartHeaders(options) {
  const headers = {};
  if (options?.flowControl) {
    const { flowControlKey, flowControlValue } = prepareFlowControl(options.flowControl);
    headers["Upstash-Flow-Control-Key"] = flowControlKey;
    headers["Upstash-Flow-Control-Value"] = flowControlValue;
  }
  if (options?.retries !== void 0) {
    headers["Upstash-Retries"] = options.retries.toString();
  }
  return headers;
}
var DLQ = class {
  constructor(client) {
    this.client = client;
  }
  /**
   * list the items in the DLQ
   *
   * @param parameters - Optional parameters object
   * @param parameters.cursor - Optional cursor for pagination
   * @param parameters.count - Optional number of items to return
   * @param parameters.filter - Optional filter options to apply to the DLQ items.
   *    The available filter options are:
   *    - `fromDate`: Filter items which entered the DLQ after this date.
   *    - `toDate`: Filter items which entered the DLQ before this date.
   *    - `url`: Filter items by the URL they were sent to.
   *    - `responseStatus`: Filter items by the response status code.
   *    - `workflowRunId`: Filter items by workflow run ID.
   *    - `workflowCreatedAt`: Filter items by workflow creation time.
   *    - `failureFunctionState`: Filter items by failure callback state.
   *    - `label`: Filter items by label.
   */
  async list(parameters) {
    const { cursor, count, filter } = parameters || {};
    return normalizeCursor(
      await this.client.http.request({
        path: ["v2", "dlq"],
        method: "GET",
        query: {
          cursor,
          count,
          ...filter,
          source: "workflow"
        }
      })
    );
  }
  async resume(request, options) {
    if (typeof request === "object" && !Array.isArray(request) && "dlqId" in request) {
      const { dlqId, flowControl, retries } = request;
      const dlqIds = Array.isArray(dlqId) ? dlqId : [dlqId];
      const { workflowRuns } = await this.client.http.request({
        path: ["v2", "workflows", "dlq", "resume"],
        query: { dlqIds },
        method: "POST",
        headers: buildResumeRestartHeaders({ flowControl, retries })
      });
      return Array.isArray(dlqId) ? workflowRuns : workflowRuns[0];
    }
    if (typeof request === "string") request = [request];
    if (Array.isArray(request) && request.length === 0) return { workflowRuns: [] };
    const filters = Array.isArray(request) ? { dlqIds: request } : request;
    return normalizeCursor(
      await this.client.http.request({
        path: ["v2", "workflows", "dlq", "resume"],
        query: buildBulkActionQueryParameters(filters),
        method: "POST",
        headers: buildResumeRestartHeaders(options)
      })
    );
  }
  async restart(request, options) {
    if (typeof request === "object" && !Array.isArray(request) && "dlqId" in request) {
      const { dlqId, flowControl, retries } = request;
      const dlqIds = Array.isArray(dlqId) ? dlqId : [dlqId];
      const { workflowRuns } = await this.client.http.request({
        path: ["v2", "workflows", "dlq", "restart"],
        query: { dlqIds },
        method: "POST",
        headers: buildResumeRestartHeaders({ flowControl, retries })
      });
      return Array.isArray(dlqId) ? workflowRuns : workflowRuns[0];
    }
    if (typeof request === "string") request = [request];
    if (Array.isArray(request) && request.length === 0) return { workflowRuns: [] };
    const filters = Array.isArray(request) ? { dlqIds: request } : request;
    return normalizeCursor(
      await this.client.http.request({
        path: ["v2", "workflows", "dlq", "restart"],
        query: buildBulkActionQueryParameters(filters),
        method: "POST",
        headers: buildResumeRestartHeaders(options)
      })
    );
  }
  /**
   * Retry the failure callback of a workflow run whose failureUrl/failureFunction
   * request has failed.
   *
   * @param dlqId - The ID of the DLQ message to retry
   * @returns response with workflow run information
   */
  async retryFailureFunction({ dlqId }) {
    const response = await this.client.http.request({
      path: ["v2", "workflows", "dlq", "callback", dlqId],
      method: "POST"
    });
    return response;
  }
  /**
   * Delete DLQ messages.
   *
   * Can be called with:
   * - A single dlqId: `delete("id")`
   * - An array of dlqIds: `delete(["id1", "id2"])`
   * - A filter object: `delete({ filter: { label: "my-label", fromDate: 1640995200000 } })`
   * - To target all entries: `delete({ all: true })`
   *
   * Processes up to `count` messages per call (defaults to 100).
   * Call in a loop until cursor is undefined to process all:
   *
   * ```ts
   * let cursor: string | undefined;
   * do {
   *   const result = await client.dlq.delete({ all: true, count: 100, cursor });
   *   cursor = result.cursor;
   * } while (cursor);
   * ```
   */
  async delete(request) {
    if (typeof request === "string") request = [request];
    if (Array.isArray(request) && request.length === 0) return { deleted: 0 };
    const filters = Array.isArray(request) ? { dlqIds: request } : request;
    return normalizeCursor(
      await this.client.http.request({
        path: ["v2", "workflows", "dlq"],
        method: "DELETE",
        query: buildBulkActionQueryParameters(filters)
      })
    );
  }
};

// src/client/index.ts
var Client = class {
  client;
  constructor(clientConfig) {
    this.client = new QStashClient(clientConfig);
  }
  async cancel(request) {
    if (typeof request === "object" && !Array.isArray(request) && ("ids" in request || "urlStartingWith" in request)) {
      const legacy = request;
      if (legacy.ids) {
        const ids = typeof legacy.ids === "string" ? [legacy.ids] : legacy.ids;
        return this.cancel(ids);
      }
      if (legacy.urlStartingWith) {
        return this.cancel({ filter: { workflowUrlStartingWith: legacy.urlStartingWith } });
      }
    }
    if (typeof request === "string") request = [request];
    if (Array.isArray(request) && request.length === 0) return { cancelled: 0 };
    const filters = Array.isArray(request) ? { workflowRunIds: request } : request;
    return await this.client.http.request({
      path: ["v2", "workflows", "runs"],
      method: "DELETE",
      query: buildBulkActionQueryParameters(filters, { translateWorkflowUrl: true })
    });
  }
  /**
   * Notify a workflow run waiting for an event
   *
   * ```ts
   * import { Client } from "@upstash/workflow";
   *
   * const client = new Client({ token: "<QSTASH_TOKEN>" })
   * await client.notify({
   *   eventId: "my-event-id",
   *   eventData: "my-data" // data passed to the workflow run
   * });
   * ```
   *
   * Optionally, you can pass a workflowRunId to enable lookback functionality:
   *
   * ```ts
   * await client.notify({
   *   eventId: "my-event-id",
   *   eventData: "my-data",
   *   workflowRunId: "wfr_123" // enables lookback
   * });
   * ```
   *
   * @param eventId event id to notify
   * @param eventData data to provide to the workflow
   * @param workflowRunId optional workflow run id for lookback support
   */
  async notify({
    eventId,
    eventData,
    workflowRunId
  }) {
    return await makeNotifyRequest(this.client.http, eventId, eventData, workflowRunId);
  }
  /**
   * Check waiters of an event
   *
   * ```ts
   * import { Client } from "@upstash/workflow";
   *
   * const client = new Client({ token: "<QSTASH_TOKEN>" })
   * const result = await client.getWaiters({
   *   eventId: "my-event-id"
   * })
   * ```
   *
   * @param eventId event id to check
   */
  async getWaiters({ eventId }) {
    return await makeGetWaitersRequest(this.client.http, eventId);
  }
  async trigger(params) {
    const isBatchInput = Array.isArray(params);
    const options = isBatchInput ? params : [params];
    const invocations = options.map((option) => {
      const failureUrl = option.failureUrl ?? option.url;
      const finalWorkflowRunId = getWorkflowRunId(option.workflowRunId);
      const context = new WorkflowContext({
        qstashClient: this.client,
        headers: new Headers({
          ...option.headers ?? {},
          ...option.label ? { [WORKFLOW_LABEL_HEADER]: option.label } : {}
        }),
        initialPayload: option.body,
        steps: [],
        url: option.url,
        workflowRunId: finalWorkflowRunId,
        telemetry: option.disableTelemetry ? void 0 : { sdk: SDK_TELEMETRY },
        label: option.label,
        workflowRunCreatedAt: Date.now()
        // pass a timestamp (server will override it)
      });
      return {
        workflowContext: context,
        telemetry: option.disableTelemetry ? void 0 : { sdk: SDK_TELEMETRY },
        delay: option.delay,
        notBefore: option.notBefore,
        failureUrl,
        retries: option.retries,
        retryDelay: option.retryDelay,
        flowControl: option.flowControl,
        redact: option.redact
      };
    });
    const result = await triggerFirstInvocation(invocations);
    const workflowRunIds = invocations.map(
      (invocation) => invocation.workflowContext.workflowRunId
    );
    if (result.isOk()) {
      return isBatchInput ? workflowRunIds.map((id) => ({ workflowRunId: id })) : { workflowRunId: workflowRunIds[0] };
    } else {
      throw result.error;
    }
  }
  /**
   * Fetches logs for workflow runs.
   *
   * @param workflowRunId - The ID of the workflow run to fetch logs for.
   * @param cursor - The cursor for pagination.
   * @param count - Number of runs to fetch. Default value is 10.
   * @param state - The state of the workflow run.
   * @param workflowUrl - The URL of the workflow. Should be an exact match.
   * @param workflowCreatedAt - The creation time of the workflow. If you have two workflow runs with the same URL, you can use this to filter them.
   * @returns A promise that resolves to either a `WorkflowRunLog` or a `WorkflowRunResponse`.
   *
   * @example
   * Fetch logs for a specific workflow run:
   * ```typescript
   * const { runs } = await client.logs({ workflowRunId: '12345' });
   * const steps = runs[0].steps; // access steps
   * ```
   *
   * @example
   * Fetch logs with pagination:
   * ```typescript
   * const { runs, cursor } = await client.logs();
   * const steps = runs[0].steps // access steps
   *
   * const { runs: nextRuns, cursor: nextCursor } = await client.logs({ cursor, count: 2 });
   * ```
   */
  async logs(params) {
    const { cursor, count, filter, ...legacyFilter } = params ?? {};
    return await this.client.http.request({
      path: ["v2", "workflows", "events"],
      query: {
        groupBy: "workflowRunId",
        ...legacyFilter,
        cursor,
        count,
        ...filter
      }
    });
  }
  get dlq() {
    return new DLQ(this.client);
  }
};
export {
  Client,
  StepTypes,
  WorkflowAbort,
  WorkflowContext,
  WorkflowError,
  WorkflowMiddleware,
  WorkflowNonRetryableError,
  WorkflowRetryAfterError,
  loggingMiddleware,
  serve
};
