// @upstash/workflow/svelte.mjs
import {
  SDK_TELEMETRY,
  serveBase,
  serveManyBase
} from "./chunk-G24J5PCC.mjs";

// platforms/svelte.ts
var telemetry = {
  sdk: SDK_TELEMETRY,
  framework: "svelte"
};
var serve = (routeFunction, options) => {
  const handler = async ({ request }) => {
    const { handler: serveHandler } = serveBase(
      routeFunction,
      telemetry,
      options,
      {
        useJSONContent: true
      }
    );
    return await serveHandler(request);
  };
  return { POST: handler };
};
var createWorkflow = (...params) => {
  const [routeFunction, options = {}] = params;
  return {
    workflowId: void 0,
    routeFunction,
    options,
    useJSONContent: true
  };
};
var serveMany = (workflows, options) => {
  return {
    POST: serveManyBase({
      workflows,
      getUrl(params) {
        return params.url.toString();
      },
      options,
      serveMethod: (routeFunction, serveOptions) => {
        return serve(routeFunction, serveOptions).POST;
      }
    }).handler
  };
};
export {
  createWorkflow,
  serve,
  serveMany
};
